Add pot and nncf tests

LeonidBeynenson opened the pull request:
Also refactor the test structure in test_ote_training.py.

LeonidBeynenson committed:
Add POT tests to test_ote_training.py

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/add_template_for_pot_nncf_tests

LeonidBeynenson committed:
Refactor test_ote_training.py -- split OTETrainingImpl into stages

LeonidBeynenson committed:
Enable POT tests in test_ote_training.py

LeonidBeynenson committed:
Remove the previous OTETrainingImpl class from test_ote_training.py

LeonidBeynenson committed:
Refactor test_ote_training.py again -- now test stages are run one-by-one automatically

LeonidBeynenson committed:
Decorative changes in test_ote_training.py

LeonidBeynenson committed:
Make one more refactor of test_ote_training.py: move evaluation to common function

LeonidBeynenson committed:
Add NNCF stages to test_ote_training.py

LeonidBeynenson committed:
Fix constraints file for the current OTE SDK

This constraints file allows avoiding unrequired uninstallation of
packages during work of the script init_venv.sh.

LeonidBeynenson committed:
Change names of final metrics in test_ote_training.py

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/add_template_for_pot_nncf_tests

AlexanderDokuchaev commented:
@LeonidBeynenson 
> should self.nncf_task.optimize be run on the whole dataset or a part of the dataset? (and which part? training? validation?)

Full dataset, like it's done in current version.

> should self.nncf_task.save_model be called?

save_model called in optimize function. Yes, no need extra call in test.

> does the code
self.nncf_model.set_data('weights.pth', trained_model.get_data('weights.pth'))
do what is required?
I mean does it loads the trained state correctly?
(If not -- NNCF will be run as training from scratch)

It's works, but if you known more correct way to set trained weights to model entity, feel free to change.

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -30,18 +30,28 @@
 from ote_sdk.entities.inference_parameters import InferenceParameters
 from ote_sdk.entities.label_schema import LabelSchemaEntity
 from ote_sdk.entities.metrics import Performance, ScoreMetric
-from ote_sdk.entities.model import ModelEntity, ModelStatus
```
This file is too large and should be splitted to several files like (utils, base, train, pot, nncf).

LeonidBeynenson replied:
I agree, but I'd like to keep this code in one file for a while (during this week and the next week).
One of the next steps will be moving the code that should be common for all backends to a special folder `test_helpers` in OTE SDK.
When I make this, the code will be splitted in a modular way.

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+        logger.info('Begin evaluation of pot model')
+        validation_dataset_pot = dataset.get_subset(self.subset)
+        score_name, score_value = run_evaluation(validation_dataset_pot, openvino_task_pot, optimized_model_pot)
+        data_collector.log_final_metric('evaluation_accuracy_pot/' + score_name, score_value)
```
I have not found comparison metric after optimization with original.

LeonidBeynenson replied:
Yes, it will be done in reallife tests implementation. It is not in the scope of this PR -- it is just the next step.

At the moment the tests run training/evaluation/etc on small datasets only just to check code integrity and that the models are runnable -- the tests are called "integration tests". So, metric comparison is not required for them.

AlexanderDokuchaev replied:
ok

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+
+        logger.info('Run NNCF optimization')
+        self.nncf_task.optimize(OptimizationType.NNCF,
+                                dataset, # TODO(lbeynens 4 adokucha): full dataset or subset here?
```
Full

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Done

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+                                self.nncf_model,
+                                OptimizationParameters())
+        logger.info('NNCF saving the model')
+        self.nncf_task.save_model(self.nncf_model) # TODO(lbeynens 4 adokucha): is it really required?
```
No, it can be removed

LeonidBeynenson replied:
\Ok

LeonidBeynenson replied:
Ok

LeonidBeynenson replied:
Done

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+            performance_improvement={},
+            model_size_reduction=1.,
+            model_status=ModelStatus.NOT_READY)
+        self.nncf_model.set_data('weights.pth', trained_model.get_data('weights.pth'))
```
I'm not sure but looks like i done it no correctyl. May be better use `self.environment_for_nncf.model = trained_model`. 
And self.nncf_model only for argument for optimize function.

LeonidBeynenson replied:
Alexander, I'm sorry, but I'm really sure that it is _not_ a good idea to use 
`self.environment_for_nncf.model = trained_model`.

The reason is as follows: the NNCF training could cause change of the trained model, so if we use directly the result of the training stage, then, if some other stages will take the result of training stage later, the stages will receive the changed model.
I would prefer to have the result of trained stage clean for any order of tests' execution.

So, I think we need to load the result of training into a new (NNCF) model. 
So, if this operation
`        self.nncf_model.set_data('weights.pth', trained_model.get_data('weights.pth'))`
loads the weights correctly, it makes what is required.
Unfortunately, I did not have time to check if it loads the weights correctly. Could you check it?

AlexanderDokuchaev replied:
>Unfortunately, I did not have time to check if it loads the weights correctly. Could you check it? 
Loading weights works correctly, checked.

LeonidBeynenson replied:
Ok

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+        logger.info('Create NNCF Task')
+        nncf_task_class_impl_path = model_template.entrypoints.nncf
+        if not nncf_task_class_impl_path:
+            pytest.skip('NNCF is not enabled for this template')
```
What do you think about add argument 'expect_nncf' in test parameters to check that nncf is configured.

LeonidBeynenson replied:
Do you mean addition pytest parameters `--expect-nncf` (like in the function `pytest_addoption` in the file `tests/conftest.py`)?

Sorry, it's not clear for me: what should we do if `--expect-nncf` is not set? just skip NNCF tests?
And what should we do if `--expect-nncf` is set but NNCF is not installed? just crash with an error?

AlexanderDokuchaev replied:
No, it's about parameters like (model_name, nncf_expected=False) - `('gen3_mobilenetV2_SSD', True)`. 
And `if nncf_expected and not nncf_task_class_impl_path`  failed test with 'nncf optimization was removed'.

LeonidBeynenson replied:
Hmmm... After some consideration with myself: why it should be added as a separate parameter?
The template file and the `test_bunches` field of the class `TestOTEIntegration` are in the same folder.
(Template files are in the subfolder `configs/ote/`, `test_bunches`  field in `tests/test_ote_training.py`)

Why do you suppose that somebody will change template file and remove the field of the template that is already in it?
It seems that it should be decided by code review stage of PR-s, not by additional parameters of tests.

AlexanderDokuchaev replied:
yes, you're right.

alexsu52 commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+class OTETestStage:
+    """
+    OTETestStage -- auxiliary class that
+    1. Allows to set up dependency between test stages: before the main action of a test stage is run, all the actions
```
Did you consider using the pytest mechanisms `@pytest.mark.dependency` to set dependencies between tests?

LeonidBeynenson replied:
I remember that I checked it and did not used it.
Now I investigated it one more time in details and has to say that, unfortunately, it does _not_ do what is required for our purpose.

Particularly, if we use `pytest.dependency` to set that `test_export` depends on `test_train`, then, when we call pytest with only `test_export` in command line arguments, the main action of `test_train` is not run.
The dependency works only for skipping tests: if `test_export` is called but `test_train` was not, then `test_export` is just skipped.

My classes work in another way: even if the only one test with `ACTION=pot_evaluation` is called through command line aerguments-- as an example this one
```
tests/test_ote_training.py::TestOTEIntegration::test[ACTION=pot_evaluation,model=gen3_mobilenetV2_ATSS,dataset=dataset1_tiled_shortened_500_A,num_iters=1,batch=2,usecase=precommit]
```
then all the required previous stages are called.

Note that for `ACTION=pot_evaluation` the previous stages are `training` and `pot` 
(and other stages that normally could be run between them, such as `training_evaluation`, `export`, `export_evaluation`, are not called, since they are not required)

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+                                            depends_stages=[pot_stage])
+        nncf_stage = OTETestStage(action=OTETestNNCFAction(),
+                                  depends_stages=[training_stage])
+        nncf_evaluation_stage = OTETestStage(action=OTETestNNCFEvaluationAction(),
```
Could you add nncf_export_stage and nncf_export_evaluation_stage. It looks like you can use OTETestExportAction and OTETestExportEvaluationAction to reuse _run_ote_export and _run_ote_export_evaluation. Need only change kwargs in __call__.

LeonidBeynenson replied:
Ok, will do it.

LeonidBeynenson replied:
Done. Please, review it..

AlexanderDokuchaev commented on tests/test_ote_training.py:
```diff
@@ -256,140 +272,453 @@ def _run_ote_training(self):
+            dataset.get_subset(self.pot_subset),
+            self.optimized_model_pot,
+            OptimizationParameters())
+        logger.info('POT optimization is finished')
```
For all optimize and train function need and check 
```
self.assertEqual(output_model.model_status, ModelStatus.SUCCESS)
```
For train it was done in previous test but missed for POT and NNCF

LeonidBeynenson replied:
I added assert-s to the code of the tests. Please, see the commit f2223ee0.

Alexander, please, check if more assert-s are required. If yes, propose your changes.

LeonidBeynenson committed:
Add more assert-s in test_ote_training.py

LeonidBeynenson committed:
Add nncf_export and nncf_export_evaluation stages to test_ote_training.py

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -30,18 +30,29 @@
+from mmdet.integration.nncf.utils import is_nncf_enabled
 
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG) #TODO(lbeynens): REMOVE BEFORE MERGE!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
```
Don't forget about this one.

LeonidBeynenson replied:
Ok

druzhkov-paul replied:
@LeonidBeynenson The last one.

druzhkov-paul replied:
```suggestion
```

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -158,31 +169,31 @@ def _convert(p):
+
+    @classmethod
+    def get_name(cls):
+        return cls._name
```
What for?

LeonidBeynenson replied:
To get the name of the class in read-only manner.

LeonidBeynenson replied:
After some consideration with myself -- made it a property.
See the commit 62f2aa0f.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -158,31 +169,31 @@ def _convert(p):
+
+    def __call__(self, data_collector: DataCollector,
+                 results_prev_stages: Optional[OrderedDict]=None):
+        raise NotImplementedError('The main action method is not implemented')
```
So, does that mean, the class and the method are abstract?

LeonidBeynenson replied:
Yes. Should I do it abstract by the proper metaclass?

LeonidBeynenson replied:
Made it an abstract class using `ABC`.
See the commit 62f2aa0f.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +272,538 @@ def _run_ote_training(self):
+            precision=[ModelPrecision.INT8],
+            target_device=TargetDevice.CPU,
+            performance_improvement={},
+            model_size_reduction=1.,
```
I guess this should be set inside the `optimize` method, right? @AlexanderDokuchaev

AlexanderDokuchaev replied:
Yes, it will be set / override inside the optimize function

LeonidBeynenson replied:
Hmmmm.... Ok what should I do here?
Could you propose your changes?

LeonidBeynenson replied:
I checked and found out that we cannot remove these attributes for the POT either.
If I remove them, the assert on the `optimization_type` field is failed:
```
>       assert self.optimized_model_pot.optimization_type == ModelOptimizationType.POT, 'Wrong optimization type'
E       AssertionError: Wrong optimization type
E       assert <ModelOptimizationType.NONE: 1> == <ModelOptimizationType.POT: 4>
E         +<ModelOptimizationType.NONE: 1>
E         -<ModelOptimizationType.POT: 4>
```
Is it expected, like for NNCF? Will it be fixed?

druzhkov-paul replied:
Please merge target branch.

LeonidBeynenson replied:
Merged. Now POT tests work.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +272,538 @@ def _run_ote_training(self):
+    _name = 'nncf'
+
+    def __init__(self, pot_subset=Subset.TRAINING):
+        self.pot_subset = pot_subset
```
Why don't you skip those "pot", "nncf", etc. prefixes inside the corresponding classes? Otherwise you'll end up having situations like this.

LeonidBeynenson replied:
Ok.

LeonidBeynenson replied:
Hmmm.... no, this `pot_subset` has another name intensionally -- since in all other cases subset is used to make valudation (so, it is `Subset.VALIDATION` by default), but there the subset is used to make POT optimization (and it is `Subset.TRAINING` by default).
So, I think, it has a special meaning here.

BTW, question for @AlexanderDokuchaev  -- is it OK if it is `Subset.TRAINING` as the default value here?

LeonidBeynenson replied:
Oh, I see -- this is copy-paste to NNCF part of code, and it has the name `pot_subset`.
**EDITED** I found out that this field is not required for NNCF at all! (at least, at the moment)

So, I remove it.

LeonidBeynenson replied:
Removed -- see the commit 7dc20629

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +272,538 @@ def _run_ote_training(self):
+            precision=[ModelPrecision.INT8],
+            target_device=TargetDevice.CPU,
+            performance_improvement={},
+            model_size_reduction=1.,
```
Remove? @AlexanderDokuchaev

AlexanderDokuchaev replied:
yes, it's useless here.

LeonidBeynenson replied:
Sorry, what should be removed?
As far as I can see, the model should be created here. Otherwise, what should be passed to the `optimize` method's call just below?
So, the creation of `ModelEntity` is required here.
Or you mean that many of these parameters in the constructor of `ModelEntiny` should be skipped?
If yes, could you propose changes?
@AlexanderDokuchaev ?

AlexanderDokuchaev replied:
It's realy can be 
```Python
ModelEntity(
            dataset,
            environment.get_model_configuration(),
            model_status=ModelStatus.NOT_READY
)
```
Some attributes will be set after https://github.com/openvinotoolkit/mmdetection/pull/206, another not used in tests.

LeonidBeynenson replied:
I checked -- no, it seems that we cannot just remove these attributes.
When I remove it then the assert on the model's `optimization_type` after optimization is failed:
```
>       assert self.nncf_model.optimization_type == ModelOptimizationType.NNCF, 'Wrong optimization type'
E       AssertionError: Wrong optimization type
E       assert <ModelOptimizationType.NONE: 1> == <ModelOptimizationType.NNCF: 3>
E         +<ModelOptimizationType.NONE: 1>
E         -<ModelOptimizationType.NNCF: 3>
```
Is it an expected behavior?
Will it be fixed?

druzhkov-paul replied:
Please merge target branch.

LeonidBeynenson replied:
I merged my target branch, but now another issue had appeared: AccuracyAwareRunner have not found mAP target metric.
See error in the comment below.

LeonidBeynenson replied:
**Update**: I found the reason of the issue and fixed the issue "AccuracyAwareRunner cannot find the metric". 
See the commit ffd893b7.

druzhkov-paul reviewed (commented):
Holly cow... This is so much code, that I didn't manage to review all of it in a single run. So, here is the review for the first half of it.

LeonidBeynenson committed:
Remove unrequired code in test_ote_training.py

LeonidBeynenson committed:
Improve the class BaseOTETestAction

Make the class BaseOTETestAction be abstract, also make property `name`
in the class instead of the method `get_name`.

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/add_template_for_pot_nncf_tests

LeonidBeynenson committed:
Remove unrequired ModelEntity fields in test_ote_training.py

LeonidBeynenson committed:
Debug insertions into EvalHook

LeonidBeynenson committed:
Fix small issue with zero metric in AccuracyAwareRunner

LeonidBeynenson committed:
Revert "Debug insertions into EvalHook"

This reverts commit a4b49f04abfe7341c4a27ac8fd4b420663479f3c.

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/add_template_for_pot_nncf_tests

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +274,521 @@ def _run_ote_training(self):
+            time the stage is called the exception is re-raised.
+    """
+    def __init__(self, action: BaseOTETestAction,
+                 depends_stages: Optional[list]=None):
```
```suggestion
                 depends_stages: Optional[List[OTETestStage]]=None):
```

druzhkov-paul commented on constraints.txt:
```diff
@@ -19,3 +19,6 @@ flatbuffers~=1.12.0
 grpcio-health-checking==1.32.0
 six~=1.15.0
+
+importlib-metadata<4.3
```
What is it used for?

LeonidBeynenson replied:
Sorry, for me it is hard to say at the moment why it is required, but once after merging your branch `dp/nous_sdk_api` to my branch when I tried to make new virtual environment I found that this line is required -- otherwise `importlib-metadata` was reinstalled during venv setup.

(I do not like reinstallation of the packages during venv setup -- sometimes this cause tricky issues with package building)

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +274,521 @@ def _run_ote_training(self):
+            return
+
+        logger.warning(f'In stage {self.name}: found that previous call of the stage '
+                       'caused exception -- re-raising it')
```
I'd suggest to go w/o `was_ok` method. The following flow looks much clearer:
```python
if not self.was_processed:
    ...
elif self.stored_exception is None:
    ...
else:
    ...
```

LeonidBeynenson replied:
Ok, will fix it right now

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +274,521 @@ def _run_ote_training(self):
+            logger.debug(f'For test stage "{self.name}": After running dep. stage "{dep_stage.name}"')
+
+        if self.was_processed:
+            self.check_is_ok()
```
This is confusing. Why do you put all checks inside `check_is_ok` if you don't want them to be checked all at once?

LeonidBeynenson replied:
I think the main problem here is naming of the function.
I think I will rename this function to `_reraise_stage_error_if_required` (and make it "protected"). 
This should make the code more clear.

druzhkov-paul replied:
And please consider removing the branch for no error case from there.

druzhkov-paul commented on tests/test_ote_training.py:
```diff
@@ -255,141 +274,521 @@ def _run_ote_training(self):
+        self.test_results_storage = OrderedDict()
+
+    def run_stage(self, stage_name, data_collector):
+        assert stage_name in self._TEST_STAGES, f'Wrong stage_name {stage_name}'
```
Don't see any reasons to have `self._TEST_STAGES` at all.

LeonidBeynenson replied:
Unfortunately, it is really required.
Please, note that there is a classmethod `OTEIntegrationTestCase.get_list_of_test_stages` that returns the list of all test stages.
The method is called in the static method `TestOTEIntegration._get_list_of_test_stages`
Whereas info from this method is required to generate the list of tests in the classmethod `TestOTEIntegration.get_list_of_tests`

Note that the method `TestOTEIntegration.get_list_of_tests` is used in the pytest magic function `pytest_generate_tests` -- this pytest trick allows to generate the list of tests before the first instantiation of tests.
Since we need the list of test stages before instantiation of any test class, we need to have a class field with the list of the test stages.

So `_TEST_STAGES` is required.
And, to be sure that we do not forget to add something (and test list generator makes its work correctly) I added some asserts with it.
Indeed, the most important assert is this one (in the method `OTEIntegrationTestCase.__init__`)
```
          assert list(self._stages.keys()) == list(self._TEST_STAGES)
```

druzhkov-paul replied:
Ok, I see. I think that the problem (the part of it actually) is that you are wrapping a simple constant with the endless number of functions/methods and then use them inconsistently. This being said, let's mark this particular comment as resolved.

LeonidBeynenson committed:
Decorative change tests/test_ote_training.py

Co-authored-by: Pavel Druzhkov <druzhkov.paul@gmail.com>

LeonidBeynenson committed:
Decorative changes in test_ote_training.py

LeonidBeynenson committed:
Merge branch 'dp/nous_sdk_api' into lb/add_template_for_pot_nncf_tests

LeonidBeynenson committed:
Fix small bug in the last decorative change in test_ote_training.py

druzhkov-paul committed:
unset debug log level
